### PR TITLE
Add projections

### PR DESCRIPTION
**Pull request by the-mikedavis**
Projections are a system for creating replicated ETS caches available on all members of a Khepri cluster. A new function `khepri_projection:new/3` creates a projection resource:

```erl
ProjectionName = wood_stocks,
ProjectionFun = fun([stock, wood, Kind], Stock) -> {Kind, Stock} end,
Options = #{type => set, read_concurrency => true},
Projection = khepri_projection:new(ProjectionName, ProjectionFun, Options).
```

The projection resource may then be registered to a pattern of nodes with `khepri:register_projection/4`:

```erl
StoreId = stock,
PathPattern = "/:stock/:wood/*",
RegisterOptions = #{},
khepri:register_projection(
  StoreId, PathPattern, Projection, RegisterOptions).
```

Once registered, the projection's ETS table is created and then filled with any existing records matching the pattern. Paths and data are passed through the `ProjectionFun` to create records which are then stored in the table. Any future changes to records which match the `PathPattern` are also applied to the projection table.

The projection table is a named ETS table which may be queried directly with `ets`:

```erl
khepri:put(StoreId, "/:stock/:wood/oak", 100),
ets:lookup_element(ProjectionName, <<"oak">>, 2).
%%=> 100
```

Projections should be used to maximize query throughput and/or minimize query latency at the cost of consistency and memory consumption. Projections have the same consistency guarantees as queries tried with the `#{favor => low_latency}` option. Data stored in the projection table is at least partially duplicated between the store and the projection table, so increased memory consumption is expected for projections with path patterns matching many nodes.

### CODE REVIEW

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:77]: I find it a bit confusing that the payload goes through several forms between the state machine and the final call to the projection function.

What do you think of using a node properties map (see `khepri_machine:gather_node_props()`) to transport the old and new payloads all to the way to the call of the projection function? This way, you can distinguish between a stored `undefined` atom and the absence of data payload.

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:147]: You could still query the arity of the extracted function later. I wonder if it's necessary to wrap the function in a `{simpl | extended, StandaloneFun}` tuple. Perhaps it's more future proof.

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:234]: See my comment earlier. In the rest of the library, we use the node properties map as the structure to carry the node details around, even if it's only the payload.

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:270]: As far as I understand, `?NO_PAYLOAD` is never passed as it is already converted to `undefined`.

File: src/khepri_tx_adv.erl
Comment (dumbbell) [inline src/khepri_tx_adv.erl:45]: Thanks! I committed this fix in `main` to not pollute this patch.

**Review by dumbbell — CHANGES_REQUESTED** [review]
I'm so glad about this addition! This is really cool :-)

I submitted several comments. I could split my feedback into two topics:
* Some documentation comments
* The way payloads are passed from the state machine to the projection function

About the payloads, I suggest we do like the rest of the library: call `khepri_machine:gather_node_props()` to create a node properties map and pass this common format to `khepri_projection:trigger()`, then decide in that function how to call the projection function.

I think we could pass the entire node properties maps (old and new) to the extended projection function, and reduce payloads to bare data for the simple one (possibly `undefined` for the old data in case a payload is added). The simple function will never be called when the new payload has no data.

What do you think?

File: src/khepri.erl
Comment (the-mikedavis) [inline src/khepri.erl:2908]: I'll push this one to master separately as well since it's unrelated to the change

File: rebar.lock
Comment (the-mikedavis) [inline rebar.lock:4]: 2.4.1 is released now as well (it bumps seshat). I can bump from 2.4.0->2.4.1 after this is merged since the commit that bumps to 2.4.0 has details on what in 2.4.0 is necessary for this feature

**Comment by the-mikedavis** [comment]
Thanks for the feedback @dumbbell!

I like the idea of passing `khepri:node_props()` around rather than payloads :+1:. I switched to that and now `khepri_projection:extended_projection_fun()`s take full node-props maps too.

For the `{simple | extended, Fun}` part: originally I thought we could avoid using standalone funs for simple projections (we can't because the funs are persisted to the log) and that tag is a relic from that. I switched to just holding on to a standalone fun and checking the arity.

File: rebar.lock
Comment (dumbbell) [reply rebar.lock:4]: Ok

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:271]: There is tiny window during which the record is missing from the ETS table. This is nearly impossible that a read happens during this window, therefore it will happen often :-)

For simple projection functions, I wonder if we should have stricter conditions on the ETS table configuration and use. Like, bags are disallowed and the key must remain the same for the same input path? This way, we don't have to remove the old record. What do you think?

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:89]: If I understand correctly, projection functions run in the context of the Ra server process.

We should document that and specify that the function should be simple and fast if possible.

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:78]: We could specify that the return value of the extended projection function is ignored.

File: src/khepri_projection.erl
Comment (dumbbell) [inline src/khepri_projection.erl:248]: I think the try/catch block should wrap the various calls to `khepri_fun:exec()` only. Otherwise it will mask crashes in Khepri itself.

**Review by dumbbell — CHANGES_REQUESTED** [review]
Thanks for the changes, it looks good to me. I have a few more tiny comments on the documentation and one question for the simple projection function.

File: src/khepri_projection.erl
Comment (the-mikedavis) [reply src/khepri_projection.erl:46]: For this one I meant "Erlang node" so I clarified that

File: src/khepri_machine.erl
Comment (the-mikedavis) [reply src/khepri_machine.erl]: I put these two into a macro so they're reused 👍

File: src/khepri_projection.erl
Comment (dumbbell) [reply src/khepri_projection.erl]: Yes, I agree, documenting this constraint is good enough.

**Review by dumbbell — APPROVED** [review]

File: src/khepri_projection.erl
Comment (dumbbell) [reply src/khepri_projection.erl]: Oops, yes, my mistake.

**Pull request merged**
